### DISCUSSION TITLE
Some fields have been changed by the VariableListener, but they are not shadow variables

### DISCUSSION THREAD

**Discussion by kentbill** [TimefoldAI/timefold-solver#1102] [General]
In the Project Job Scheduling example of QuickStarts, the fields **startDate**, **endDate**, and **busyDates** have been changed by the VariableListener, as shown in the following code.

[https://github.com/TimefoldAI/timefold-quickstarts/blob/68b4e3f05954c04a97459390416b11887acfa3bf/java/project-job-scheduling/src/main/java/org/acme/projectjobschedule/domain/solver/PredecessorsDoneDateUpdatingVariableListener.java#L47](url)

and

[https://github.com/TimefoldAI/timefold-quickstarts/blob/68b4e3f05954c04a97459390416b11887acfa3bf/java/project-job-scheduling/src/main/java/org/acme/projectjobschedule/domain/Allocation.java#L172](url)

Does it violate the principle that "**VariballeListener can only modify shadow variables**"?
[https://github.com/TimefoldAI/timefold-solver/blob/c411ceba8fa58bccdd93e9bd83a889eb9e3ccbc5/docs/src/modules/ROOT/pages/using-timefold-solver/modeling-planning-problems.adoc?plain=1#L872](url)

**Comment by triceo** [comment]
Hello @kentbill! Well spotted.

Variable listeners can only modify shadow variables, as opposed to genuine variables. That was supposed to have been the key message there. 

In the Quickstart, you see us modifying fields which are neither genuine nor shadow variables. This is not a problem, as long as:

- No other variable depends on those fields. If it did, the listeners wouldn't trigger for the dependent variable, resulting in score corruptions.
- At least one genuine and/or shadow variable is changed on every entity which contains those fields that are being changed. If nothing was updated on the entity, yet the entity still changed, this would result in score corruption.

Even though our use in the Quickstart is safe, it is clearly a grey area and we should've avoided it. We are prototyping on a new approach to shadow variables which - if it works out - will eliminate these issues by eliminating variable listeners. ;-) Stay tuned!

**Reply by kentbill** [reply]
Thank you for your reply! @triceo.
Looking forward to the new method of Shadow Variable, it sounds great!
I noticed a new annotation in version 1.13.0 - **@CascadingUpdateShadowVariable**. Is this one of the new features of the Shadow mutable approach?
﻿

**Reply by triceo** [reply]
It does solve many issues with the variable listeners, and it is the recommended way of updating tail chains for now.
